### DISCUSSION TITLE
Exported Application Stuck Loading

### DISCUSSION THREAD

**Discussion by VacationTV** [rrirower/DP2SceneGraphWizard#59] [Q&A]
Thanks for your work on this, I feel pretty close but need some help. 

I've built a package from my JSON file (https://www.vacationchannels.com/wp-content/uploads/2023/12/TheVacationChannel-120623.json), and I see the logo, splash screen, and then get stuck with "Please wait. Loading..."

I have telnet open and the below is the crash code I am receiving. Can anyone help point me in the right direction with this information? Many thanks! -Martin

Suspending threads...
Thread selected:  1*   ...ents/tasks/MainLoaderTask.brs(615)   functionQueryObjectById(id as string, contentType as object) as Object

Current Function:
615:* function QueryObjectById(id as string, contentType as object) as Object
616:      object = invalid
617:      rexp = Substitute("\b{0}\b", id)
618:      regxp = CreateObject("roRegex", rexp, "i")
619:      ' Parse movies.
Source Digest(s): 
pkg:            dev  1.0.0 bcbb1343 The Vacation Channel
roku_ads_lib: 58484 3.7.11 7823d89b Roku Ads Library

Type Mismatch. Unable to cast "roArray" to "String". (runtime error &h18) in pkg:/components/tasks/MainLoaderTask.brs(615)
Backtrace:
#2  Function queryobjectbyid(id As String, contenttype As Object) As Object
   file/line: pkg:/components/tasks/MainLoaderTask.brs(615)
#1  Function parsejsonfeed(feedurl As String) As Void
   file/line: pkg:/components/tasks/MainLoaderTask.brs(152)
#0  Function getcontent() As Void
   file/line: pkg:/components/tasks/MainLoaderTask.brs(281)
Local Variables:
id               roArray refcnt=3 count:4
contenttype      roArray refcnt=2 count:1
global           Interface:ifGlobal
m                roAssociativeArray refcnt=4 count:4
object           <uninitialized>
rexp             <uninitialized>
regxp            <uninitialized>
movie            <uninitialized>
livefeed         <uninitialized>
shortformvideo   <uninitialized>
tvspecial        <uninitialized>
Threads:
ID    Location                                Source Code
 0    pkg:/source/main.brs(74)                       msg = wait(0, m.eventmessages)
 1*   ...ents/tasks/MainLoaderTask.brs(615)   functionQueryObjectById(id as string, contentType as object) as Object
  *selected

**Comment by rrirower** [comment]
At first glance, it looks like you have a problem with your feed file. You can use the Channel Feed Manager software to validate your feed file.

I'll try to take a closer look when I have more time.

**Reply by VacationTV** [reply]
I sincerely appreciate it. I should have also said that I had no issues with the videos in the feed file loading with Roku’s conversion sdk files, but they all lumped into a ShortForm row. 

I dug into the code and realised I was going to have to modify to parse categories. That lead me to your code and it seemed like a great solution.

I left work woth a plan to start my morning trying out your feed manager. Thanks!

**Comment by rrirower** [comment]
Your feed file is malformed.  You have syntax errors in the file.  For example, your playlist objects have an extra set of brackets.

Please review the [Roku Json Spec](https://developer.roku.com/docs/specs/direct-publisher-feed-specs/json-dp-spec.md#playlist) for more info.

**Reply by VacationTV** [reply]
Thanks for your help. Your tools are really fantastic. I'm trying to sort out as many errors on my own as possible, but I have hit a wall where I'd appreciate your advice. Many thanks in advance for staying so involved on this project. 

I've loaded the feed into your Channel Manager and got a 0 error validation. 

I uploaded that feed here: https://www.vacationchannels.com/wp-content/uploads/2023/12/TheVacationChannel-120823.json

I re-ran DP2SceneGraph and sent the  package to my box. I am still stuck on a "Please Wait. Loading.." message. This is the telnet code I am seeing: 


------ Running dev 'The Vacation Channel' main ------
12-07 18:27:47.302 [beacon.signal] |AppLaunchComplete ---------> Pending Render Pass
12-07 18:27:47.349 [beacon.signal] |AppLaunchComplete ---------> Duration(1308 ms)

BrightScript Micro Debugger.
Enter any BrightScript statement, debug commands, or HELP.

Suspending threads...
Thread selected:  1*   ...ents/tasks/MainLoaderTask.brs(154)                                  row = QueryChannelRowByCategoryName(category.name, rootChildren)

Current Function:
146:              for each playlist in m.jsonContentFeed.playlists
147:                  category = GetCategoryFromPlaylist(playlist.name)
148:                  if category <> invalid
149:                      if playlist.itemIds <> invalid
150:                          for each id in playlist.itemIds
151:                              playlistObjectType[0] = invalid
152:                              playlistObject = QueryObjectById(id, playlistObjectType)
153:                              if playlistObject <> invalid
154:*                                 row = QueryChannelRowByCategoryName(category.name, rootChildren)
155:                                  ' Build assoc array for the row.
156:                                  'itemData = GetJsonItemData(playlistObject, playlistObjectType[0])
157:                                  itemData = GetJsonItemData(playlistObject, "playlist")
158:                                  row.children.Push(itemData)
Source Digest(s): 
pkg:            dev  1.0.0 02e86690 The Vacation Channel
roku_ads_lib: 58484 3.7.11 7823d89b Roku Ads Library

Syntax Error. (runtime error &h02) in pkg:/components/tasks/MainLoaderTask.brs(154)
Backtrace:
#1  Function parsejsonfeed(feedurl As String) As Void
   file/line: pkg:/components/tasks/MainLoaderTask.brs(154)
#0  Function getcontent() As Void
   file/line: pkg:/components/tasks/MainLoaderTask.brs(281)
Local Variables:
feedurl          roString (2.1 was String) refcnt=2 val:"https://www.vacationchannels.com/wp-content/uploads/2023/12/TheVacationChannel-120823.json"
global           Interface:ifGlobal
m                roAssociativeArray refcnt=3 count:4
xfer             roUrlTransfer refcnt=1
rsp              roString (2.1 was String) refcnt=1 val:"{
  "lastUpdated": "2023-12-07T16:46:23+00:00",
  "providerName": "The Vacation Channel",
 "...
rootchildren     roArray refcnt=1 count:20
category         roString refcnt=1 val:""
row              roAssociativeArray refcnt=2 count:2
movie            <uninitialized>
categories       roArray refcnt=1 count:0
categoryname     <uninitialized>
itemdata         <uninitialized>
livefeed         <uninitialized>
series           <uninitialized>
seasons          <uninitialized>
shortformvideo   roAssociativeArray refcnt=2 count:10
tvspecial        <uninitialized>
playlistobjecttype roArray refcnt=1 count:1
playlist         roAssociativeArray refcnt=2 count:2
id               roString (2.1 was String) refcnt=2 val:"26"
playlistobject   roAssociativeArray refcnt=2 count:10
contentnode      <uninitialized>
Threads:
ID    Location                                Source Code
 0    pkg:/source/main.brs(74)                       msg = wait(0, m.eventmessages)
 1*   ...ents/tasks/MainLoaderTask.brs(154)                                  row = QueryChannelRowByCategoryName(category.name, rootChildren)
  *selected

Brightscript Debugger>

**Reply by rrirower** [reply]
@VacationTV It looks like you tried to combine Categories defined by 'tags' with Categories defined by playlists.  In any event, you have a playlist called 'Smithfield' that is not associated with any Category, so, the code 'hung'.

I will fix the code to ignore playlists that are not associated with a Category.  This problem also alerted me to an error in the validation code for the Channel Feed Manager.  That code should have reported the problem to you earlier.  I will fix that also.

**Reply by VacationTV** [reply]
Thank you so much for the attention. I will work it through on my end. I've got 4 channels that I will be converting and I look forward to moving my workflow onto your tools in the future. 

Is there a PayPal account (or similar) I can use to buy you a beer? Cheers!

**Reply by rrirower** [reply]
@VacationTV 

`Is there a PayPal account (or similar) I can use to buy you a beer?`

Not required, but, appreciated.  Hit the 'About' button on the first page of the Wizard.
